Remotion/Web/Development.Analyzers/RMWEB0001_WebStringAnalyzer.cs

```diff
@@ -0,0 +1,223 @@
+        if (SymbolEqualityComparer.Default.Equals (argumentType, symbols.WebStringTypeSymbol)
+            || SymbolEqualityComparer.Default.Equals (argumentType, symbols.PlainTextStringTypeSymbol))
+        {
+          var diagnostic = Diagnostic.Create (
+              DiagnosticDescriptor,
+              ctx.Node.GetLocation(),
+              symbols.HtmlTextWriterWriteMethodSymbol.ToString(),
+              argumentType.ToString(),
+              $"Use '{argumentType.ToString()}.Write(HtmlTextWriter)' instead.");
+          ctx.ReportDiagnostic (diagnostic);
+        }
+      }
+      else if (AnyWithResult (
+          symbols.StringJoinMethodSymbols,
+          s => SymbolEqualityComparer.Default.Equals (invocationOriginalDefinitionSymbol, s),
+          out var methodSymbol))
+      {
+        var argumentTypes = GetArgumentTypeSymbols (node, ctx.SemanticModel);
+        if (AnyWithResult (
+            argumentTypes,
+            a => SymbolEqualityComparer.Default.Equals (a, symbols.WebStringTypeSymbol)
+                 || SymbolEqualityComparer.Default.Equals (a, symbols.PlainTextStringTypeSymbol)
+                 || ctx.Compilation.ClassifyConversion (a, symbols.WebStringEnumerable).IsImplicit
+                 || ctx.Compilation.ClassifyConversion (a, symbols.PlainTextStringEnumerable).IsImplicit,
+            out var result))
+        {
+          var diagnostic = Diagnostic.Create (
+              DiagnosticDescriptor,
+              ctx.Node.GetLocation(),
+              methodSymbol.ToString(),
+              result.ToString(),
+              $"Encode the {result.ToString()} instances first.");
+          ctx.ReportDiagnostic (diagnostic);
+        }
+      }
+    }
+
+    private static bool AnyWithResult<TSymbol> (IEnumerable<TSymbol> methods, Func<TSymbol, bool> predicate, out TSymbol match)
+        where TSymbol : ISymbol
+    {
+      match = methods.FirstOrDefault (predicate)!;
+      return match != null;
+    }
+
+    private static ITypeSymbol? GetFirstArgumentTypeSymbol (InvocationExpressionSyntax node, SemanticModel model)
+    {
+      var firstArgumentExpression = node.ArgumentList.Arguments.First().Expression;
+      return model.GetTypeInfo (firstArgumentExpression).Type;
+    }
+
+    private static IReadOnlyCollection<ITypeSymbol> GetArgumentTypeSymbols (InvocationExpressionSyntax node, SemanticModel model)
+    {
+      return node.ArgumentList.Arguments.Select (s => model.GetTypeInfo (s.Expression).Type).OfType<ITypeSymbol>().ToList();
+    }
+
+    private static IReadOnlyCollection<IMethodSymbol> GetStringJoinMethodSymbols (INamedTypeSymbol stringSymbol)
+    {
+      return stringSymbol
+          .GetMembers ("Join")
+          .Cast<IMethodSymbol>()
+          .ToList();
+    }
+
+    private static IMethodSymbol GetHtmlTextWriterWriteObjectMethodSymbol (INamedTypeSymbol htmlTextWriterSymbol, INamedTypeSymbol? objectSymbol)
```

Comment:
Why are some of the parameters nullable and others not? Is this just derived from the usage in the comparer? The callsite doesn't seem to indicate that we need them as nullable.